### PR TITLE
Add the random infections feature #272

### LINKED ISSUE #272: Add the random infections feature
POC: Omer Deutsch.

This is a long to explain issue relatively not too complex to implement.

I created a full explanation here -
https://docs.google.com/document/d/11_wLMJCeRzqOEkZbSPU7QIRawnpae4jsQ6aa-BnCuvU/edit#

This is super important for the project currently. Notify me that you're starting this, and feel free to ask me questions, I hope the document is self-explanatory.

### PR DESCRIPTION
**Pull request by meirhalachmi**

### CODE REVIEW

File: src/corona_hakab_model/generation/circles_generator.py [hunk 1]
```diff
@@ -104,8 +109,37 @@ def __init__(
...
+                    # No random connections
+                    continue
+
+                agents_id = [a.index for a in circle.agents]
+
+                # Sample from exponential distribution
+                rand_connections = np.random.exponential(exp_mean, len(agents_id))
+
+                # You can't have more random connections than the number of people (other than you) in the circle
+                rand_connections[rand_connections > num_of_agents_in_circle - 1] = num_of_agents_in_circle - 1
```
Comment (bgmoshe) [inline src/corona_hakab_model/generation/circles_generator.py]: Depending on the size of the array, np.clip() would prob. be better (but, it does create a new array so...)

Comment (meirhalachmi) [reply src/corona_hakab_model/generation/circles_generator.py]: forgot about `np.clip()`. thx.

File: src/corona_hakab_model/generation/circles_generator.py [hunk 2]
```diff
@@ -104,8 +111,37 @@ def __init__(
...
+                    continue
+
+                agents_id = [a.index for a in circle.agents]
+
+                # Sample from exponential distribution
+                rand_connections = np.random.exponential(exp_mean, len(agents_id))
+
+                # You can't have more random connections than the number of people (other than you) in the circle
+                # rand_connections = np.clip(rand_connections, a_max=num_of_agents_in_circle - 1)  # TODO: Should we clip?
+                circle.total_random_connections = sum(rand_connections)
```
Comment (tal-nc) [inline src/corona_hakab_model/generation/circles_generator.py]: maybe np.sum would be faster?

File: src/corona_hakab_model/update_matrix.py [hunk 3]
```diff
@@ -83,10 +96,19 @@ def change_connections_policy(self, connection_types_to_use: Iterable[Connection
         self.matrix.set_factors(factors)
         self.normalize()
 
+    def reset_agent(self, connection_type, index):
+        self.matrix.reset_mul_row(connection_type, index)
+        self.matrix.reset_mul_col(connection_type, index)
+        self.manager.random_connections_factor[index, connection_type] = 1  # TODO: Make sure that's the meaning of "reset"
```
Comment (tal-nc) [inline src/corona_hakab_model/update_matrix.py]: that's the meaning. you can remove the TODO

**Review by tal-nc — COMMENTED** [review]
looks great so far. 
the main thing missing is to split connection types to those who only meet random people from the same social circle, and those who meet random people by geographical circle.

as we already said, moving this to matrix_data would be nice, but can wait untill the first version is released.

File: src/corona_hakab_model/infection.py [hunk 4]
```diff
@@ -42,14 +48,31 @@ def _perform_infection(self) -> List[Agent]:
...
 
-        # new_infected: dict -
-        # key = medical state (currently only susceptible state which an agent can be infected)
-        # value = list of agents
-        new_infected = list()
-        for index in infected_indices:
-            agent = self.manager.agents[index]
-            new_infected.append(agent)
+                probs_not_infected_from_connection[(agents_id, [connection_type] * len(agents_id))] = \
+                    1 - prob * self.manager.random_connections_strength[connection_type]
```
Comment (bgmoshe) [inline src/corona_hakab_model/infection.py:72]: You do not need the 1 here, - is good enough :)

File: src/corona_hakab_model/update_matrix.py [hunk 5]
```diff
@@ -69,8 +75,15 @@ def normalize(self):
             # saves this for the effective r0 graph
             self.total_contagious_probability = total_contagious_probability
 
+            # Random connections
+            random_total = np.dot(self.manager.num_of_random_connections.sum(0),
```
Comment (bgmoshe) [inline src/corona_hakab_model/update_matrix.py:79]: Why 0?

File: src/corona_hakab_model/update_matrix.py [hunk 6]
```diff
@@ -69,8 +75,15 @@ def normalize(self):
...
+                                  self.manager.random_connections_strength)
+
             # this factor should be calculated once when the matrix is full, and be left un-changed for the rest of the run.
-            self.normalize_factor = (beta * self.size) / (self.matrix.total())
+            self.normalize_factor = (beta * self.size) / (self.matrix.total() + random_total)
+
+            # Normalize random connections (only once!)
+            self.manager.random_connections_strength *= self.normalize_factor
```
Comment (bgmoshe) [inline src/corona_hakab_model/update_matrix.py:86]: I am OK with it, but can u please explain  to me why it is needed?

**Review by bgmoshe — APPROVED** [review]
Some small holes, but we can merge this in my opinion

File: src/corona_hakab_model/infection.py (hunk 4 above)
Comment (meirhalachmi) [reply src/corona_hakab_model/infection.py:72]: what? why?

File: src/corona_hakab_model/update_matrix.py (hunk 5 above)
Comment (meirhalachmi) [reply src/corona_hakab_model/update_matrix.py:79]: axis=0

Comment (meirhalachmi) [reply src/corona_hakab_model/update_matrix.py:79]: sum on all agents

File: src/corona_hakab_model/update_matrix.py (hunk 6 above)
Comment (tal-nc) [reply src/corona_hakab_model/update_matrix.py:86]: for the same reason normalizing the matrix is important. 
we want to make sure that the average amount of people a sick agent will infect will be R0. that includes both the people he regularly meets, and those he randomly meets.

**Pull request merged**